### PR TITLE
✨ Link previews

### PR DESCRIPTION
**Pull request by jozsefsallai**
TODO:
- [x] Link preview component
- [x] Dynamic link preview using a content proxy
- [x] Workaround for CORS issues that happen when accessing contentproxy.okuna.io (still quite wonky)
- [x] Mobile optimization (looks fine but not fine enough)

This PR adds a new required parameter in the dotenv files: `CONTENTPROXY_URL`. For local developments, set it to `/contentproxy`.

### CODE REVIEW

**Comment by flogruber** [comment]
Link previews on the web sign as title and subtitle Okuna, Ethical Social Network.  
Is this behaviour planned?  

**web-app**
image

**mobile-app**
temp

**Comment by lifenautjoe** [comment]
cc @evict, if you have some time to take a look together with https://github.com/OkunaOrg/okuna-web/pull/86.

 Basically, we do the same as the mobile app. Fetch the website through the proxy, crawl it for preview image and favicon, retrieve those items through the proxy again but then convert them into base64 and insert them as 

```js
        get imageUrl() {
            if (this.preview.imageData) return "data:" + this.preview.imageContentType + ";base64," + this.preview.imageData;
        }

        get faviconUrl() {
            if (this.preview.faviconData) return "data:" + this.preview.faviconContentType + ";base64," + this.preview.faviconData;
        }
```


```html
                           <figure class="image is-16x16 is-semi-rounded">
                                    <img :src="faviconUrl"/>
                                </figure>
```

Where faviconContentType we also check its of valid image type, otherwise we dont preview it.

```js
            if(contentType !== 'image/png' && contentType !== 'image/jpeg' && contentType !== 'image/gif') return;
```

File: services/link-preview/LinkPreviewService.ts [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            if (images && images.length) {
+                for (const image in images) {
+                    const imageElement = images[image];
+                    const imageSrc: string = imageElement?.attributes?.['src']?.value;
+
+                    if (!imageSrc) {
+                        continue;
+                    }
+
+                    if (imageSrc.endsWith('jpg') || imageSrc.endsWith('gif') || imageSrc.endsWith('png')) {
```
Comment (evict) [inline services/link-preview/LinkPreviewService.ts:122]: Maybe add SVG too.

File: components/post/components/post-link-preview/lib/GetFirstLink.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import { IPost } from '~/models/posts/post/IPost';
+import UrlMatcher from '~/lib/matchers/UrlMatcher';
+
+function isSafeURL(url: string): boolean {
+    try {
+        const parsedURL = new URL(url);
+        const forbiddenURLs = ['localhost', '127.0.0.1', '0.0.0.0'];
```
Comment (evict) [inline components/post/components/post-link-preview/lib/GetFirstLink.ts:7]: Ipv6 also has localhost on: `::1` & `0000:0000:0000:0000:0000:0000:0000:0001` and 0.0.0.0 is `::/128`.

**Comment by lifenautjoe** [comment]
@evict Also good to know, we changed using DOMParser for another client side library https://github.com/OkunaOrg/okuna-web/commit/46cf37cd8890201d61fd7781209381e902c61dbf

File: services/utils/UtilsService.ts [hunk 3]
```diff
@@ -157,19 +157,45 @@ export class UtilsService implements IUtilsService {
...
         const regexp = /(ftp|http|https):\/\/(\w+:{0,1}\w*@)?(\S+)(:[0-9]+)?(\/|\/([\w#!:.?+=&%@!\-\/]))?/;
         return regexp.test(str);
     }
+
+
+    normalizeUrl({url, rootUrl}: NormalizeUrlParams) : string {
+
+        let normalizedUrl;
+
+        if (url.startsWith('http') || url.startsWith('http')) {
```
Comment (evict) [inline services/utils/UtilsService.ts:181]: This statement makes no sense. I think you can remove `|| url.startsWith('http')`?

File: services/utils/UtilsService.ts [hunk 4]
```diff
@@ -157,19 +157,45 @@ export class UtilsService implements IUtilsService {
...
         let res = {};
-        (new URL(query)).searchParams.forEach((value: string, key: string,)=>{
+        (new URL(query)).searchParams.forEach((value: string, key: string,) => {
             res[key] = value;
         });
         return res;
     };
 
-    isUrl(str: string) : boolean {
+    isUrl(str: string): boolean {
```
Comment (evict) [inline services/utils/UtilsService.ts:171]: Any reason why we have FTP in there?

**Comment by evict** [comment]
Looks good! I was happy when I saw `InnerText`.

File: components/post/components/post-link-preview/lib/GetFirstLink.ts (hunk 2 above)
Comment (jozsefsallai) [reply components/post/components/post-link-preview/lib/GetFirstLink.ts:7]: Nice catch! Completely forgot about ipv6 :)

File: services/link-preview/LinkPreviewService.ts (hunk 1 above)
Comment (jozsefsallai) [reply services/link-preview/LinkPreviewService.ts:122]: I thought about this but I wasn't entirely sure if SVGs would be safe enough. But as long as they're inside of an img tag and base64 encoded it shouldn't be an issue I think.

File: services/utils/UtilsService.ts (hunk 3 above)
Comment (jozsefsallai) [reply services/utils/UtilsService.ts:181]: Probably a typo. Judging from the rest of the code, this was probably meant to be `url.startsWith('https')`

Comment (lifenautjoe) [reply services/utils/UtilsService.ts:181]: **Yep!**

**Comment by jozsefsallai** [comment]
all done in 8f8db17475d92e311d5abc98f466a5527710d619 :)

**Comment by evict** [comment]
Alright! But the statement would be true for both http and https so you do not need the second statement anyway!

**Pull request merged**